File: tools/export-routes.lua
-- Developer-only export; run from the catalog root with lua5.4.
local output = assert(arg[1], "usage: lua5.4 tools/export-routes.lua output.tsv")
local routes = {}
local points = {}
local starts = dofile("tools/route-starts.lua")
local function point(value)
  return type(value) == "table" and type(value.x) == "number"
    and type(value.y) == "number" and type(value.plane) == "number"
end
local function collect_points(prefix, values)
  for name, value in pairs(values or {}) do
    if point(value) then points[prefix .. "." .. name] = value end
  end
end
local function add_route(name, route, within)
  local first = assert(route.start or starts[name], "missing audit start for " .. name)
  local tiles = { first }
  for _, via in ipairs(route.via or {}) do tiles[#tiles + 1] = via end
  tiles[#tiles + 1] = assert(route.destination, "missing route destination " .. name)
  assert(point(tiles[#tiles]), "invalid route " .. name)
  routes[name] = { points = tiles, within = route.within or within,
    arrival_tiles = route.arrival_tiles or {}, avoid_tiles = route.avoid_tiles or {},
    account = route.account or "ALL" }
end
local configs = assert(io.popen("rg --files scripts/quest-runner -g config.lua"))
local paths = {}
for path in configs:lines() do paths[#paths + 1] = path end
assert(configs:close())
table.sort(paths)
for _, path in ipairs(paths) do
  local name = assert(path:match("quest%-runner/([^/]+)/config.lua$"))
  local config = dofile(path)
  collect_points(name, config.points)
  for route_name, route in pairs(config.routes or {}) do
    add_route(name .. "." .. route_name, route, 2)
  end
  if config.points and config.points.maze_inside then
    add_route(name .. ".maze_route", { destination = config.points.maze_inside }, 0)
  end
end
for name, route in pairs(dofile("tools/transport-routes.lua")) do add_route(name, route) end
local stream = assert(io.open(output, "w"))
stream:write("# name\tindex\tx\ty\tplane\twithin\tkind\taccount\n")
local names = {}
for name in pairs(routes) do names[#names + 1] = name end
table.sort(names)
for _, name in ipairs(names) do
  local route = routes[name]
  for index, tile in ipairs(route.points) do
    stream:write(string.format("%s\t%d\t%d\t%d\t%d\t%d\troute\t%s\n",
      name, index, tile.x, tile.y, tile.plane, route.within, route.account))
  end
  for index, tile in ipairs(route.arrival_tiles) do
    assert(point(tile), "invalid arrival tile " .. name)
    stream:write(string.format("%s\t%d\t%d\t%d\t%d\t%d\tarrival\t%s\n",
      name, index, tile.x, tile.y, tile.plane, route.within, route.account))
  end
  for index, tile in ipairs(route.avoid_tiles) do
    assert(point(tile), "invalid avoid tile " .. name)
    stream:write(string.format("%s\t%d\t%d\t%d\t%d\t%d\tavoid\t%s\n",
      name, index, tile.x, tile.y, tile.plane, route.within, route.account))
  end
end
local point_names = {}
for name in pairs(points) do point_names[#point_names + 1] = name end
table.sort(point_names)
for _, name in ipairs(point_names) do
  local tile = points[name]
  stream:write(string.format("%s\t1\t%d\t%d\t%d\t0\tpoint\tALL\n", name, tile.x, tile.y, tile.plane))
end
assert(stream:close())
print(string.format("Exported %d routes and %d declared points", #names, #point_names))

File: src/routeAudit/java/com/genericclient/scripts/quests/ExportRoutes.java
package com.genericclient.scripts.quests;

import com.genericclient.script.Navigation.Journey;
import java.io.BufferedWriter;
import java.io.IOException;
import java.lang.reflect.Field;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import org.dreambot.api.methods.map.Tile;

/** Export real catalog geometry; starts describe the script's calling locations. */
public final class ExportRoutes
{
    private ExportRoutes() {}

    public static void main(String[] args) throws Exception
    {
        Path output=Path.of(args[0]);
        Files.createDirectories(output.toAbsolutePath().getParent());
        Map<String,Tile> starts=monkeyStarts();
        try (BufferedWriter writer=Files.newBufferedWriter(output))
        {
            writer.write("# name\tindex\tx\ty\tplane\twithin\tkind\taccount\n");
            Field[] routes=MonkeyRoutes.class.getDeclaredFields();
            Arrays.sort(routes,Comparator.comparing(Field::getName));
            for (Field field : routes)
            {
                if (field.getType()!=Journey.class) continue;
                Journey journey=(Journey)field.get(null);
                String name=field.getName().toLowerCase(Locale.ROOT);
                Tile start=starts.remove(name);
                if (start==null) throw new IllegalArgumentException("Missing caller start: " + name);
                route(writer,"monkey_madness_i."+name,start,journey,"ALL");
                if (name.startsWith("prison_to_") || name.equals("temple_to_monkey_child"))
                    rows(writer,"monkey_madness_i."+name,prisonTiles(),0,"forbidden","ALL");
            }
            if (!starts.isEmpty()) throw new IllegalArgumentException("Unused caller starts: " + starts.keySet());
            route(writer,"tree_gnome_village.maze_inside",new Tile(2505,3190),new Journey(new Tile(2515,3159),0),"ALL");
            route(writer,"tree_gnome_village.maze_exit",new Tile(2515,3159),new Journey(new Tile(2505,3190),0),"ALL");
            route(writer,"waterfall.gnome_dungeon",new Tile(2440,3089),new Journey(WaterfallNavigation.GNOME_BASEMENT,1),"QUEST_ROUTES");
            route(writer,"grand_tree.stronghold",new Tile(2440,3089),new Journey(GnomeTravel.KING_TILE,2),"ALL");
            route(writer,"grand_tree.hazelmere",new Tile(2440,3089),new Journey(GnomeTravel.HAZELMERE,1),"ALL");
            transports(writer);
        }
        System.out.println("Exported Java catalog journeys to " + output);
    }

    private static void transports(BufferedWriter writer) throws IOException
    {
        route(writer,"witchs_house.basement",new Tile(2902,3473),new Journey(WitchsHouse.BASEMENT_ENTRY,0),"ALL");
        route(writer,"witchs_house.surface",new Tile(2901,9874),new Journey(WitchsHouse.SURFACE_ENTRY,0),"ALL");
        route(writer,"waterfall.raft",new Tile(2510,3493),new Journey(WaterfallNavigation.HUDON_LANDING,1),"QUEST_ROUTES");
        route(writer,"waterfall.tourist_upstairs",new Tile(2519,3430),new Journey(WaterfallNavigation.TOURIST_UPSTAIRS,0),"ALL");
        route(writer,"waterfall.tourist_ground",new Tile(2518,3431,1),new Journey(WaterfallNavigation.TOURIST_GROUND,0),"ALL");
        route(writer,"waterfall.gnome_basement",new Tile(2448,3090),new Journey(WaterfallNavigation.GNOME_BASEMENT,1),"QUEST_ROUTES");
        route(writer,"waterfall.gnome_surface",new Tile(2548,9566),new Journey(WaterfallNavigation.GNOME_SURFACE,1),"QUEST_ROUTES");
        route(writer,"waterfall.tomb_exit",new Tile(2542,9813),new Journey(WaterfallNavigation.TOMB_SURFACE,1),"ALL");
        route(writer,"the_grand_tree.hazelmere",new Tile(2677,3088),new Journey(GnomeTravel.HAZELMERE,1),"ALL");
        route(writer,"the_grand_tree.glough",new Tile(2466,3494),new Journey(GnomeTravel.GLOUGH_ROOM,2),"ALL");
        route(writer,"the_grand_tree.return_from_glough",new Tile(2477,3463,1),new Journey(GnomeTravel.KING_TILE,2),"ALL");
        route(writer,"the_grand_tree.charlie",new Tile(2466,3494),new Journey(GnomeTravel.TOP,1),"ALL");
        route(writer,"the_grand_tree.anita",new Tile(2466,3494,3),new Journey(GnomeTravel.ANITA,1),"ALL");
        route(writer,"the_grand_tree.invasion_plans",new Tile(2388,3513,1),new Journey(GnomeTravel.GLOUGH_ROOM,2),"ALL");
        route(writer,"monkey_madness_i.spirit_tree",new Tile(3184,3508),new Journey(MonkeyMap.STRONGHOLD_ARRIVAL,1),"QUEST_ROUTES");
        route(writer,"monkey_madness_i.shipyard",new Tile(2466,3494),new Journey(MonkeyMap.SHIPYARD_GATE,3),"QUEST_ROUTES");
        route(writer,"monkey_madness_i.return_from_gandius",new Tile(2970,2972),new Journey(MonkeyMap.KING_NARNODE,3),"QUEST_ROUTES");
        route(writer,"monkey_madness_i.daero",new Tile(2466,3494,3),new Journey(MonkeyMap.DAERO,5),"ALL");
        route(writer,"monkey_madness_i.repeat_hangar",new Tile(2461,3444),new Journey(MonkeyMap.POST_PUZZLE_LANDING,1),"QUEST_ROUTES");
        route(writer,"monkey_madness_i.waydar",new Tile(2649,4516),new Journey(MonkeyMap.CRASH_ISLAND_LANDING,1),"QUEST_ROUTES");
        route(writer,"monkey_madness_i.lumdo",new Tile(2894,2726),new Journey(MonkeyMap.APE_ATOLL_LANDING,1),"QUEST_ROUTES");
    }

    private static Map<String,Tile> monkeyStarts()
    {
        Map<String,Tile> starts=new LinkedHashMap<>();
        starts.put("ape_atoll_valley",new Tile(2770,2707));
        starts.put("denture_safe_approach",new Tile(2764,2764));
        starts.put("garkor_to_dentures",new Tile(2807,2762));
        starts.put("garkor_to_west_ladder",new Tile(2775,2767));
        starts.put("marim_gate_to_garkor",new Tile(2738,2767));
        starts.put("prison_to_dentures",new Tile(2762,2804));
        starts.put("prison_to_garkor",new Tile(2762,2804));
        starts.put("prison_to_temple_entry",new Tile(2762,2804));
        starts.put("temple_to_monkey_child",new Tile(2806,2785));
        starts.put("temple_trapdoor_approach",new Tile(2787,2787));
        starts.put("zoo_to_grand_tree",new Tile(2608,3278));
        starts.put("zooknock_dungeon",new Tile(2768,9101));
        return starts;
    }

    private static List<Tile> prisonTiles()
    {
        List<Tile> tiles=new ArrayList<>();
        for (Map<String,Integer> bounds : MonkeyAreas.PRISON_BOUNDS)
            for (int x=bounds.get("x1");x<=bounds.get("x2");x++)
                for (int y=bounds.get("y1");y<=bounds.get("y2");y++)
                    tiles.add(new Tile(x,y,bounds.get("plane")));
        return tiles;
    }

    private static void route(BufferedWriter writer, String name, Tile start, Journey journey, String account) throws IOException
    {
        List<Tile> route=new ArrayList<>();
        route.add(start); route.addAll(journey.via); route.add(journey.destination);
        rows(writer,name,route,journey.within,"route",account);
        rows(writer,name,journey.arrival,journey.within,"arrival",account);
        rows(writer,name,journey.avoid,journey.within,"avoid",account);
    }

    private static void rows(BufferedWriter writer, String name, List<Tile> tiles, int within, String kind, String account) throws IOException
    {
        for (int index=0; index<tiles.size(); index++)
        {
            Tile tile=tiles.get(index);
            writer.write(String.format(Locale.ROOT,"%s\t%d\t%d\t%d\t%d\t%d\t%s\t%s%n",name,index+1,tile.getX(),tile.getY(),tile.getZ(),within,kind,account));
        }
    }
}
